---
title: integration of 10x and inDrop data
description: Here we use the collated inDrop and 10x datasets and re-identify the cell types from the previous analysis
category: research
subcategory: rnaseq
tags: [singlecell_rnaseq]
Principal Investigator: Hubert Perrimon
Researcher: Ruei-Jiun Hung
Experiment: Drosophila gut single-cell RNA-seq [hbc02022]
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    theme: readable
    df_print: paged
    fig_width: 8.5
    fig_height: 5
---

```{r set-publication-theme, echo=FALSE, cache=FALSE}
ggplot2::theme_set(ggpubr::theme_pubr(base_size=10))
```

```{r setup, echo = FALSE, cache = FALSE}
knitr::opts_chunk$set(dev = c('png', 'cairo_pdf'), 
                      fig.align = 'center', fig.height = 5, fig.width = 8.5, 
                      pdf.options(encoding = "ISOLatin9.enc"),
                      fig.path='figures/', warning=FALSE, message=FALSE)
```

```{r load-libraries}
library(Seurat)
library(Matrix)
library(tidyverse)
library(viridis)
```

# Load the Seurat object
Before we load the object, we're going to switck the rownames of our data to
be the symbols. There is a 1-1 mapping of gene ID to symbol in these samples so
we can do this with no loss of information.

```{r load-data}
dat = readRDS(file.path("data", "counts.rds"))
md = readRDS(file.path("data", "metadata.rds"))
g2s = readr::read_csv(file.path("metadata", "tx2gs.csv"),
                      col_names=c("gene", "symbol", "transcript")) %>%
  dplyr::select(-transcript) %>%
  unique()
rownames(dat) = g2s$symbol[match(rownames(dat), g2s$gene)]
seurat = CreateSeuratObject(dat, meta.data=md)
```

# Overview
First we'll look at the uncorrected data; we'll just correct for mitochondrial
percentage and nothing else. Then we'll run PCA, clustering and look at the UMAP
plots to get an idea of how much run-to-run and technology variation we are 
dealing with.

```{r uncorrected-data}
seurat = SCTransform(seurat, vars.to.regress = "pctMito", verbose = FALSE)
seurat = RunPCA(seurat, verbose = FALSE)
seurat = RunUMAP(seurat, dims = 1:30, verbose = FALSE)
```

```{r technology-matters}
p1 = DimPlot(seurat, group.by = "technology")
p1
```

A lot! We can see that the 10x and the inDrop data are mostly separate from each
other, but there is some overlaps. We can see looking at the four samples that
the four samples are very different from each other as well.

```{r sample-matters}
DimPlot(seurat, group.by = "sample")
```

So we are going to have some work to do trying to overlay these on top of each
other. There are a few ways to do that. One is to just regress out the effect of
the technology, the other is to do something more complex like CCA. We do need
to do something about it though if we want to combine these two datasets. However, even
without doing anything about the technological variation we can take a peek at the data
and see if we can pick out some patterns. 

# uncorrected clustering
Here we'll do a rough clustering and then look at the expression of our markers, without correcting for
which technology or experiment they are from, to orient outselves.

## clustering
```{r jackstraw}
ElbowPlot(seurat, ndims=50)
seurat = FindNeighbors(seurat, dims=1:25)
seurat = FindClusters(seurat, resolution=0.5)
DimPlot(seurat)
```


## cell type markers 
```{r cell-type-markers}
markers = readr::read_csv("metadata/cell_type_markers.csv")
knitr::kable(markers)
```

Below we can see that cluster 3 and cluster 21 are likely ISC+EB or EB cells. Cluster 9, 17 and 20 are
liekly some subtypes of EE cells. Groups 1, 4, 11, 14 and 23 might be aEC cells. Groups 5, 13 and 
15 might be pEC cells. Group 8 looks like mECs. And group 15 looks like cardia cells.

Cluster 12 looks like it might be LFC cells. Cluster 7 might be iron or copper cells. 
Cluster 6 looks like it could be an EC cell, but we don't know what compartment.

We're just missing 2, 10, 18, 19, and 22. Not bad.

```{r average-expression}
cluster.averages = AverageExpression(seurat, return.seurat=TRUE, add.ident="sample")
DoHeatmap(cluster.averages, features=head(markers, 20)$genes, size=2)
DoHeatmap(cluster.averages, features=tail(markers, 21)$genes, size=2)
```

```{r store-rough-cell-type-calls}
clusters = seurat@meta.data$seurat_clusters
rough_celltype = case_when(
  clusters %in% c(3, 21) ~ "ISC/EB",
  clusters %in% c(9, 17, 20) ~ "EE",
  clusters %in% c(0, 1, 4, 11, 14, 23) ~ "aEC",
  clusters %in% c(5, 13, 16) ~ "pEC",
  clusters %in% c(8, 24) ~ "mEC",
  clusters %in% c(6) ~ "EC",
  clusters %in% c(15) ~ "cardia",
  clusters %in% c(12) ~ "LFC",
  clusters %in% c(7) ~ "iron/copper",
  TRUE ~ as.character(clusters))
seurat@meta.data$rough_celltype = rough_celltype
Idents(seurat) = seurat@meta.data$rough_celltype
roughcelltype.averages = AverageExpression(seurat, return.seurat=TRUE, add.ident="sample")
DoHeatmap(roughcelltype.averages, features=head(markers, 20)$genes, size=2)
DoHeatmap(roughcelltype.averages, features=tail(markers, 21)$genes, size=2)
p3 = DimPlot(seurat, group.by="rough_celltype", label=TRUE)
p3
```

It looks like we are able to identify the cells across both technologies, which is very good.

```{r rough-cell-type-counts}
knitr::kable(seurat@meta.data %>%
             group_by(rough_celltype, technology) %>%
             summarise(n=n()))
```


## unmarked clusters
This leaves us with clusters 2, 10, 18, 19 and 22 as unknown clusters.

```{r unmarked-clusters}
seurat@meta.data %>%
  filter(is.na(rough_celltype)) %>%
  pull(seurat_clusters) %>%
  unique()
```


## trypsin genes {.tabset}

```{r trypsin-genes}
tryp = c("deltaTry", "gammaTry", "alphaTry", "betaTry",
         "epsilonTry", "thetaTry", "kappaTry", "lambdaTry",
         "iotaTry", "etaTry", "zetaTry")
r1 = c("deltaTry", "gammaTry", "Muc68D")
r2 = c("deltaTry", "gammaTry", "alphaTry", "betaTry", "epsilonTry", "thetaTry")
r3 = c("thetaTry")
r4 = c("kappaTry", "lambdaTry", "iotaTry", "etaTry", "zetaTry")
r5 = c("iotaTry", "etaTry", "zetaTry")
crop = c("Cyp312a1", "Cyp4e3", "Spn27A", "spz", "pot")
compartments = unique(c(r1, r2, r3, r4, r5, crop))
FeaturePlot(seurat, features=r2)
FeaturePlot(seurat, features=r4)
FeaturePlot(seurat, features=r5)
FeaturePlot(seurat, features=crop)
DoHeatmap(cluster.averages, features=compartments, size=2)
```

It looks like group 12 and 15 might be from R1. Group 0, 1, 4, 11, 14, 19, 23 and 24 are from either R2 or R3. 
Group 7 looks like it might be from R3. Groups 5, 6 and 16 look like they are from R4. Group 18 looks
like it is from the crop.

```{r trypsin-gene-heatmap}
clusters = seurat@meta.data$seurat_clusters
compartment = case_when(
  clusters %in% c(12, 15) ~ "R1",
  clusters %in% c(0, 1, 4, 11, 14, 19, 23, 24) ~ "R2",
  clusters %in% c(7) ~ "R3",
  clusters %in% c(5, 6, 16) ~ "R4",
  clusters %in% c(18) ~ "crop",
  TRUE ~ as.character(clusters))
seurat@meta.data$compartment = compartment
Idents(seurat) = seurat@meta.data$compartment
compartment.averages = AverageExpression(seurat, return.seurat=TRUE, add.ident="sample")
DoHeatmap(compartment.averages, features=compartments, size=2)
p5 = DimPlot(seurat, group.by="compartment", label=TRUE)
p5
```

It looks like we are able to identify compartments cells are from across both technologies, which is very good.

```{r trypsin-by-technology}
knitr::kable(seurat@meta.data %>%
  group_by(compartment, technology) %>%
  summarise(n=n()))
```

## write seurat
```{r write-rough-seurat}
saveRDS(seurat, file.path("results", "seurat-rough.rds"))
```

# integration

Here we wil use the integration functions in Seurat to integrate the 10x and the inDrop datasets.

```{r subtype-samples}
Idents(seurat) = seurat@meta.data$seurat_clusters
indrop = SubsetData(seurat, subset.name="technology", accept.value="inDrop")
indrop = NormalizeData(object = indrop)
indrop = ScaleData(object = indrop)
indrop = FindVariableFeatures(indrop, do.plot = FALSE)
tenx = SubsetData(seurat, subset.name="technology", accept.value="10x")
tenx = NormalizeData(object = tenx)
tenx = ScaleData(object = tenx)
tenx = FindVariableFeatures(tenx, do.plot = FALSE)
hvf = union(VariableFeatures(indrop), VariableFeatures(tenx))
anchors = FindIntegrationAnchors(object.list=c(indrop, tenx), dims=1:30)
integrated = IntegrateData(anchorset = anchors, dims = 1:30, features.to.integrate=rownames(indrop))
DefaultAssay(integrated) = "integrated"
```

# integrated clustering
Here we use the integrated data to re-do the clustering. After integration, we can see that the 10x and
the inDrop datasets overlap with each other much better. We can also get a better idea about what
the identities of cells might be.

```{r integrated-clustering}
integrated = ScaleData(integrated, verbose = FALSE)
integrated = RunPCA(integrated, npcs = 30, verbose = FALSE)
integrated = RunUMAP(integrated, reduction = "pca", dims = 1:30)
integrated = FindNeighbors(integrated, reduction = "pca", dims = 1:30)
integrated = FindClusters(integrated, resolution = 0.5)
p2 = DimPlot(integrated, reduction = "umap", group.by = "technology")
p4 = DimPlot(integrated, reduction = "umap", group.by = "rough_celltype", label=TRUE)
p6 = DimPlot(integrated, group.by="compartment", label=TRUE)
library(gridExtra)
grid.arrange(p1 + ggtitle("before integration"), p2 + ggtitle("after integration"))
grid.arrange(p3 + ggtitle("before integration"), p4 + ggtitle("after integration"))
grid.arrange(p5 + ggtitle("before integration"), p6 + ggtitle("after integration"))
```

Here we redo the marker expression, using the integrated cluster IDs rather than the unintegrated
IDs from before.

```{r integrated-clusters}
DimPlot(integrated, group.by="seurat_clusters", label=TRUE)
integrated.averages = AverageExpression(integrated, return.seurat=TRUE, add.ident="sample")
DoHeatmap(integrated.averages, features=head(markers$genes, 20), size=2)
DoHeatmap(integrated.averages, features=tail(markers$genes, 21), size=2)
knitr::kable(markers)
```

Here we are missing classifying cluster 9, 13, and 16. classying 19 and 20 as ISC cells is pretty weak, IMO,
so maybe those too. 

```{r mark-new-clusters}
clusters = integrated@meta.data$seurat_clusters
integrated_celltype = case_when(
  clusters %in% c(3) ~ "ISC/EB",
  clusters %in% c(12, 14, 15) ~ "EE",
  clusters %in% c(1, 4, 18) ~ "aEC",
  clusters %in% c(6) ~ "iron/copper",
  clusters %in% c(5, 11, 13) ~ "pEC",
  clusters %in% c(8) ~ "mEC",
  clusters %in% c(7) ~ "LFC",
  clusters %in% c(17) ~ "cardia",
  TRUE ~ as.character(clusters))
integrated@meta.data$integrated_celltype = integrated_celltype
Idents(integrated) = integrated@meta.data$integrated_celltype
integratedcelltype.averages = AverageExpression(integrated, return.seurat=TRUE, add.ident="sample")
DoHeatmap(integratedcelltype.averages, features=markers$genes, size=2)
p7 = DimPlot(integrated, group.by="integrated_celltype", label=TRUE)
grid.arrange(p4 + ggtitle("integrated celltypes"), p7 + ggtitle("integrated celltypes"))
knitr::kable(integrated@meta.data %>%
             dplyr::select(seurat_clusters, integrated_celltype) %>%
             unique(), row.names=FALSE)
```

# cluster markers (ROC) {.tabset}
Here we will find markers for each of the clusters where we haven't been able
to assign a cell type so we can hopefully figure out what type of cells they
are.

```{r set-cluster-ids}
Idents(integrated) = integrated@meta.data$seurat_cluster
```


## Cluster 0
```{r cluster1-markers_roc}
dir.create(file.path("results", "markers_roc"))
markersroc_0 = FindMarkers(integrated, ident.1=0, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markersroc_0, file.path("results", "markers_roc", "cluster0-markers_roc.csv"))
knitr::kable(head(markersroc_0, 25))
```


## Cluster 1
```{r cluster0-markers_roc}
dir.create(file.path("results", "markers_roc"))
markers_roc1 = FindMarkers(integrated, ident.1=1, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc1, file.path("results", "markers_roc", "cluster1-markers_roc.csv"))
knitr::kable(head(markers_roc1, 25))
```

## Cluster 2
```{r cluster2-markers_roc}
dir.create(file.path("results", "markers_roc"))
markers_roc2 = FindMarkers(integrated, ident.1=2, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc2, file.path("results", "markers_roc", "cluster2-markers_roc.csv"))
knitr::kable(head(markers_roc2, 25))
```


## Cluster 3
```{r cluster3-markers_roc}
markers_roc3 = FindMarkers(integrated, ident.1=3, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc3, file.path("results", "markers_roc", "cluster3-markers_roc.csv"))
knitr::kable(head(markers_roc3, 25))
```


## Cluster 4
```{r cluster4-markers_roc}
markers_roc4 = FindMarkers(integrated, ident.1=4, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc4, file.path("results", "markers_roc", "cluster4-markers_roc.csv"))
knitr::kable(head(markers_roc4, 25))
```

## Cluster 5
```{r cluster5-markers_roc}
markers_roc5 = FindMarkers(integrated, ident.1=5, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc5, file.path("results", "markers_roc", "cluster5-markers_roc.csv"))
knitr::kable(head(markers_roc5, 25))
```


## Cluster 6
```{r cluster6-markers_roc}
markers_roc6 = FindMarkers(integrated, ident.1=6, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc6, file.path("results", "markers_roc", "cluster6-markers_roc.csv"))
knitr::kable(head(markers_roc6, 25))
```


## Cluster 7
```{r cluster7-markers_roc}
markers_roc7 = FindMarkers(integrated, ident.1=7, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc7, file.path("results", "markers_roc", "cluster7-markers_roc.csv"))
knitr::kable(head(markers_roc7, 25))
```


## Cluster 8
```{r cluster8-markers_roc}
markers_roc8 = FindMarkers(integrated, ident.1=8, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc8, file.path("results", "markers_roc", "cluster8-markers_roc.csv"))
knitr::kable(head(markers_roc8, 25))
```

## Cluster 9
```{r cluster9-markers_roc}
markers_roc9 = FindMarkers(integrated, ident.1=9, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc9, file.path("results", "markers_roc", "cluster9-markers_roc.csv"))
knitr::kable(head(markers_roc9, 25))
```


## Cluster 10
```{r cluster10-markers_roc}
markers_roc10 = FindMarkers(integrated, ident.1=10, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc10, file.path("results", "markers_roc", "cluster10-markers_roc.csv"))
knitr::kable(head(markers_roc10, 25))
```


## Cluster 11
```{r cluster11-markers_roc}
markers_roc11 = FindMarkers(integrated, ident.1=11, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc11, file.path("results", "markers_roc", "cluster11-markers_roc.csv"))
knitr::kable(head(markers_roc11, 25))
```


## Cluster 12
```{r cluster12-markers_roc}
markers_roc12 = FindMarkers(integrated, ident.1=12, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc12, file.path("results", "markers_roc", "cluster12-markers_roc.csv"))
knitr::kable(head(markers_roc12, 25))
```


## Cluster 13
```{r cluster13-markers_roc}
markers_roc13 = FindMarkers(integrated, ident.1=13, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc13, file.path("results", "markers_roc", "cluster13-markers_roc.csv"))
knitr::kable(head(markers_roc13, 25))
```


## Cluster 14
```{r cluster14-markers_roc}
markers_roc14 = FindMarkers(integrated, ident.1=14, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc14, file.path("results", "markers_roc", "cluster14-markers_roc.csv"))
knitr::kable(head(markers_roc14, 25))
```


## Cluster 15
```{r cluster15-markers_roc}
markers_roc15 = FindMarkers(integrated, ident.1=15, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc15, file.path("results", "markers_roc", "cluster15-markers_roc.csv"))
knitr::kable(head(markers_roc15, 25))
```


## Cluster 16
```{r cluser16-markers_roc}
markers_roc16 = FindMarkers(integrated, ident.1=16, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc16, file.path("results", "markers_roc", "cluster16-markers_roc.csv"))
knitr::kable(head(markers_roc16, 25))
```


## Cluster 17
```{r cluster17-markers_roc}
markers_roc17 = FindMarkers(integrated, ident.1=17, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc17, file.path("results", "markers_roc", "cluster17-markers_roc.csv"))
knitr::kable(head(markers_roc17, 25))
```

## Cluster 18
```{r cluster18-markers_roc}
markers_roc18 = FindMarkers(integrated, ident.1=18, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc18, file.path("results", "markers_roc", "cluster18-markers_roc.csv"))
knitr::kable(head(markers_roc18, 25))
```


## Cluster 19
```{r cluster19-markers_roc}
markers_roc19 = FindMarkers(integrated, ident.1=19, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc19, file.path("results", "markers_roc", "cluster19-markers_roc.csv"))
knitr::kable(head(markers_roc19, 25))
```

## Cluster 20
```{r cluster20-markers_roc}
markers_roc20 = FindMarkers(integrated, ident.1=20, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc20, file.path("results", "markers_roc", "cluster20-markers_roc.csv"))
knitr::kable(head(markers_roc20, 25))
```

## Cluster 21
```{r cluster21-markers_roc}
markers_roc21 = FindMarkers(integrated, ident.1=21, test.use="roc") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(-power)
readr::write_csv(markers_roc21, file.path("results", "markers_roc", "cluster21-markers_roc.csv"))
knitr::kable(head(markers_roc21, 25))
```

Specific markers-roc plot for cluster 21.
```{r plot-markers-roc-cluster21}
DimPlot(integrated, label=TRUE)
FeaturePlot(integrated, features=c("MtnA", "Vha13", "Smvt", "Oatp58Dc"))
cells21 = rownames(subset(integrated@meta.data, seurat_clusters == 21))
FeaturePlot(integrated[,cells21], features=c("MtnA", "Vha13", "Smvt", "Oatp58Dc")) +
  ggtitle("cluster 21 only")
```

# cluster markers (MAST) {.tabset}
Here we will find markers for each of the clusters where we haven't been able
to assign a cell type so we can hopefully figure out what type of cells they
are.


## Cluster 0
```{r cluster0-markers_MAST}
dir.create(file.path("results", "markers_MAST"))
markers_MAST0 = FindMarkers(integrated, ident.1=0, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST0, file.path("results", "markers_MAST", "cluster0-markers_MAST.csv"))
knitr::kable(head(markers_MAST0, 25))
```

## Cluster 1
```{r cluster1-markers_MAST}
dir.create(file.path("results", "markers_MAST"))
markers_MAST1 = FindMarkers(integrated, ident.1=1, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST1, file.path("results", "markers_MAST", "cluster1-markers_MAST.csv"))
knitr::kable(head(markers_MAST1, 25))
```

## Cluster 2
```{r cluster2-markers_MAST}
dir.create(file.path("results", "markers_MAST"))
markers_MAST2 = FindMarkers(integrated, ident.1=2, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST2, file.path("results", "markers_MAST", "cluster2-markers_MAST.csv"))
knitr::kable(head(markers_MAST2, 25))
```


## Cluster 3
```{r cluster3-markers_MAST}
markers_MAST3 = FindMarkers(integrated, ident.1=3, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST3, file.path("results", "markers_MAST", "cluster3-markers_MAST.csv"))
knitr::kable(head(markers_MAST3, 25))
```


## Cluster 4
```{r cluster4-markers_MAST}
markers_MAST4 = FindMarkers(integrated, ident.1=4, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST4, file.path("results", "markers_MAST", "cluster4-markers_MAST.csv"))
knitr::kable(head(markers_MAST4, 25))
```

## Cluster 5
```{r cluster5-markers_MAST}
markers_MAST5 = FindMarkers(integrated, ident.1=5, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST5, file.path("results", "markers_MAST", "cluster5-markers_MAST.csv"))
knitr::kable(head(markers_MAST5, 25))
```


## Cluster 6
```{r cluster6-markers_MAST}
markers_MAST6 = FindMarkers(integrated, ident.1=6, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST6, file.path("results", "markers_MAST", "cluster6-markers_MAST.csv"))
knitr::kable(head(markers_MAST6, 25))
```


## Cluster 7
```{r cluster7-markers_MAST}
markers_MAST7 = FindMarkers(integrated, ident.1=7, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST7, file.path("results", "markers_MAST", "cluster7-markers_MAST.csv"))
knitr::kable(head(markers_MAST7, 25))
```


## Cluster 8
```{r cluster8-markers_MAST}
markers_MAST8 = FindMarkers(integrated, ident.1=8, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST8, file.path("results", "markers_MAST", "cluster8-markers_MAST.csv"))
knitr::kable(head(markers_MAST8, 25))
```

## Cluster 9
```{r cluster9-markers_MAST}
markers_MAST9 = FindMarkers(integrated, ident.1=9, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST9, file.path("results", "markers_MAST", "cluster9-markers_MAST.csv"))
knitr::kable(head(markers_MAST9, 25))
```


## Cluster 10
```{r cluster10-markers_MAST}
markers_MAST10 = FindMarkers(integrated, ident.1=10, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST10, file.path("results", "markers_MAST", "cluster10-markers_MAST.csv"))
knitr::kable(head(markers_MAST10, 25))
```


## Cluster 11
```{r cluster11-markers_MAST}
markers_MAST11 = FindMarkers(integrated, ident.1=11, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST11, file.path("results", "markers_MAST", "cluster11-markers_MAST.csv"))
knitr::kable(head(markers_MAST11, 25))
```


## Cluster 12
```{r cluster12-markers_MAST}
markers_MAST12 = FindMarkers(integrated, ident.1=12, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST12, file.path("results", "markers_MAST", "cluster12-markers_MAST.csv"))
knitr::kable(head(markers_MAST12, 25))
```


## Cluster 13
```{r cluster13-markers_MAST}
markers_MAST13 = FindMarkers(integrated, ident.1=13, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST13, file.path("results", "markers_MAST", "cluster13-markers_MAST.csv"))
knitr::kable(head(markers_MAST13, 25))
```


## Cluster 14
```{r cluster14-markers_MAST}
markers_MAST14 = FindMarkers(integrated, ident.1=14, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST14, file.path("results", "markers_MAST", "cluster14-markers_MAST.csv"))
knitr::kable(head(markers_MAST14, 25))
```


## Cluster 15
```{r cluster15-markers_MAST}
markers_MAST15 = FindMarkers(integrated, ident.1=15, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST15, file.path("results", "markers_MAST", "cluster15-markers_MAST.csv"))
knitr::kable(head(markers_MAST15, 25))
```


## Cluster 16
```{r cluser16-markers_MAST}
markers_MAST16 = FindMarkers(integrated, ident.1=16, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST16, file.path("results", "markers_MAST", "cluster16-markers_MAST.csv"))
knitr::kable(head(markers_MAST16, 25))
```


## Cluster 17
```{r cluster17-markers_MAST}
markers_MAST17 = FindMarkers(integrated, ident.1=17, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST17, file.path("results", "markers_MAST", "cluster17-markers_MAST.csv"))
knitr::kable(head(markers_MAST17, 25))
```

## Cluster 18
```{r cluster18-markers_MAST}
markers_MAST18 = FindMarkers(integrated, ident.1=18, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST18, file.path("results", "markers_MAST", "cluster18-markers_MAST.csv"))
knitr::kable(head(markers_MAST18, 25))
```


## Cluster 19
```{r cluster19-markers_MAST}
markers_MAST19 = FindMarkers(integrated, ident.1=19, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST19, file.path("results", "markers_MAST", "cluster19-markers_MAST.csv"))
knitr::kable(head(markers_MAST19, 25))
```

## Cluster 20
```{r cluster20-markers_MAST}
markers_MAST20 = FindMarkers(integrated, ident.1=20, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST20, file.path("results", "markers_MAST", "cluster20-markers_MAST.csv"))
knitr::kable(head(markers_MAST20, 25))
```

## Cluster 21
```{r cluster21-markers_MAST}
markers_MAST21 = FindMarkers(integrated, ident.1=21, test.use="MAST") %>%
  as.data.frame() %>%
  tibble::rownames_to_column("gene") %>%
  arrange(p_val_adj)
readr::write_csv(markers_MAST21, file.path("results", "markers_MAST", "cluster21-markers_MAST.csv"))
knitr::kable(head(markers_MAST21, 25))
```


```{r save-integration}
saveRDS(integrated, file.path("results","integrated.rds"))
```

